### PR TITLE
Email verification by PIN

### PR DESCRIPTION
**Pull request by jpslav**
Provides new APIs for getting contact info and confirming (verifying) contact infos.

* Contact info information provided in `GET /api/user` (see API docs)
* `PUT /api/contact_infos/:id/resend_confirmation` (see API docs)
* `PUT /api/contact_infos/:id/confirm` (see API docs)

## Misc
* Updates `lev` to `2.2.1` to get support for `after_transaction` blocks.
* Monkey patched in ActiveRecord enum support from Rails 4.2 :-)
* Added a `to have_routine_error_code` and `to have_routine_error_status` rspec helpers.
* Added should-matchers

### CODE REVIEW

**Comment by jpslav** [comment]
@nathanstitt please review when you have a chance

File: spec/access_policies/contact_info_access_policy_spec.rb [hunk 1]
```diff
@@ -35,19 +35,4 @@
...
-                                               contact_info)).to eq true
-      expect(OSU::AccessPolicy.action_allowed?(:confirm, temp,
-                                               contact_info)).to eq true
-      expect(OSU::AccessPolicy.action_allowed?(:confirm, user,
-                                               contact_info)).to eq true
-      expect(OSU::AccessPolicy.action_allowed?(:confirm, admin,
-                                               contact_info)).to eq true
-    end
-  end
-
```
Comment (nathanstitt) [inline spec/access_policies/contact_info_access_policy_spec.rb]: Were these specs moved into the controller spec below? If not, should we have retained them?

File: lib/active_record_enum.rb [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+          }
+        elsif !klass_method && method_defined_within?(method_name, _enum_methods_module, Module)
+          raise ArgumentError, ENUM_CONFLICT_MESSAGE % {
+            enum: enum_name,
+            klass: self.name,
+            type: 'instance',
+            method: method_name,
+            source: 'another enum'
+          }
+        end
```
Comment (nathanstitt) [inline lib/active_record_enum.rb]: This is a neat use of `%` to fill in the `ENUM_CONFLICT_MESSAGE` messages.  It's really just `sprintf` but I like how it reads. I know, you just copied it in :)

File: app/routines/confirm_by_pin.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        after_transaction do
+          sequential_failure.reset!
+        end
+
+        run(ConfirmContactInfo, contact_info)
+      else
+        after_transaction do
+          sequential_failure.increment!
+        end
+
```
Comment (nathanstitt) [inline app/routines/confirm_by_pin.rb]: I'm probably missing something, but why are call `.reset!` and `.increment!` in an `after_transaction`   Doesn't that mean that those methods may fail silently without causing the lev routine to fail?  Maybe we want that behavior for some reason though.

**Comment by nathanstitt** [comment]
@jpslav looks great!  I'm guessing the `after_transaction` blocks are needed but thought I'd point them out just in case.

File: spec/access_policies/contact_info_access_policy_spec.rb (hunk 1 above)
Comment (jpslav) [inline spec/access_policies/contact_info_access_policy_spec.rb]: Confirmation by code is always allowed -- previously the access policy just returned true but it wasn't even called (the handler just said `def authorized? true; end`) so I took out the code and this spec that tested it.

File: app/routines/confirm_by_pin.rb (hunk 3 above)
Comment (jpslav) [inline app/routines/confirm_by_pin.rb]: The `sequential_failure.reset!` call should probably be inside the routine's transaction, but the `increment!` call needs to be afterwards.  The reason is that the `fatal_error` call will rollback the routine transaction which will undo the increment (which we don't want).

**Pull request merged**